src/index.tsx

```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
     if (isEqual(graph.nodes, prevNodes.current)) {
       return; // No change!
     }
-    const { added, removed, updated } = diff(prevNodes.current, graph.nodes);
+    const { added, removed, updated } = diff(nodes.get(), graph.nodes);
 
-    nodes.remove(removed);
-    nodes.add(added);
-    nodes.update(updated);
+    if (removed.length) {
+      nodes.remove(removed);
+    }
+    // Shallow clone nodes to ensure props aren't mutated!
+    if (added.length) {
+      nodes.add(shallowClone(added));
+    }
+    if (updated.length) {
+      nodes.update(shallowClone(updated));
+    }
     prevNodes.current = graph.nodes;
   }, [graph.nodes, nodes]);
 
   useEffect(() => {
     if (isEqual(graph.edges, prevEdges.current)) {
       return; // No change!
     }
-    const { added, removed, updated } = diff(prevEdges.current, graph.edges);
-
-    edges.remove(removed);
-    edges.add(added);
-    edges.update(updated);
+    const { added, removed, updated } = diff(edges.get(), graph.edges);
+    if (removed.length) {
+      edges.remove(removed);
+    }
+    // Shallow clone edges to ensure props aren't mutated!
+    if (added.length) {
+      edges.add(shallowClone(added));
```

Comment:
same as above

Reply:
The main reason I added these if checks was to avoid a bunch of no-op function calls since they were easily avoidable.